File: Eclat/Phuc/Eclat.java
package Eclat.Phuc;

import java.io.*;
import java.util.*;

public class Eclat {

    static final int MAX_TRANS = 4000; // đủ cho chess.txt

    // Đọc transactions từ file
    public static List<List<String>> readTransactions(String filename) throws IOException {
        List<List<String>> transactions = new ArrayList<>();
        BufferedReader br = new BufferedReader(new FileReader(filename));
        String line;
        while ((line = br.readLine()) != null) {
            String[] tokens = line.trim().split("\\s+");
            transactions.add(Arrays.asList(tokens));
        }
        br.close();
        return transactions;
    }

    // Chuyển transactions sang map<item, BitSet>
    public static Map<String, BitSet> convert(List<List<String>> transactions) {
        Map<String, BitSet> res = new HashMap<>();
        for (int i = 0; i < transactions.size(); i++) {
            for (String item : transactions.get(i)) {
                res.computeIfAbsent(item, k -> new BitSet(MAX_TRANS)).set(i);
            }
        }
        return res;
    }

    // Eclat với BitSet
    public static void eclat(List<Map.Entry<String, BitSet>> items, BitSet prefixTids, List<String> prefix, int minSup,
                             List<Pair<List<String>, Integer>> results, int nTrans) {
        for (int i = 0; i < items.size(); i++) {
            String item = items.get(i).getKey();
            BitSet itemTids = (BitSet) items.get(i).getValue().clone();

            BitSet newTids;
            if (prefix.isEmpty()) {
                newTids = itemTids;
            } else {
                newTids = (BitSet) prefixTids.clone();
                newTids.and(itemTids);
            }

            int support = newTids.cardinality();
            if (support >= minSup) {
                prefix.add(item);
                results.add(new Pair<>(new ArrayList<>(prefix), support));

                // Chuẩn bị các item phía sau
                List<Map.Entry<String, BitSet>> newItems = new ArrayList<>();
                for (int j = i + 1; j < items.size(); j++) {
                    newItems.add(items.get(j));
                }
                if (!newItems.isEmpty()) {
                    eclat(newItems, newTids, prefix, minSup, results, nTrans);
                }
                prefix.remove(prefix.size() - 1);
            }
        }
    }

    // Sắp xếp tăng dần theo support
    public static Comparator<Map.Entry<String, BitSet>> supportComparator() {
        return Comparator.comparingInt(e -> e.getValue().cardinality());
    }

    // Định nghĩa Pair đơn giản
    public static class Pair<A, B> {
        public A first;
        public B second;
        public Pair(A a, B b) { first = a; second = b; }
    }

    public static void main(String[] args) throws IOException {
        String filename = "chess.txt"; // Đặt đúng đường dẫn file
        int minSup = 1500; // Chỉnh theo nhu cầu

        List<List<String>> transactions = readTransactions(filename);
        int nTrans = transactions.size();

        Map<String, BitSet> itemTid = convert(transactions);

        // Lọc các item đủ minSup
        List<Map.Entry<String, BitSet>> itemSet = new ArrayList<>();
        for (Map.Entry<String, BitSet> x : itemTid.entrySet()) {
            if (x.getValue().cardinality() >= minSup) {
                itemSet.add(x);
            }
        }
        itemSet.sort(supportComparator());

        List<Pair<List<String>, Integer>> ans = new ArrayList<>();
        eclat(itemSet, new BitSet(MAX_TRANS), new ArrayList<>(), minSup, ans, nTrans);

        System.out.println("Các tập phổ biến:");
        for (Pair<List<String>, Integer> x : ans) {
            for (String y : x.first) System.out.print(y + " ");
            System.out.println(": " + x.second);
        }
    }
}

File: Eclat/Phuc/Eclat.cpp
#include <bits/stdc++.h>
#define endl "\n"

using namespace std;

map<string, set<int>> convert(vector<vector<string>> &transaction){
    map<string, set<int>> res;
    for(int i=0;i<(int)transaction.size();i++){
        for(string item : transaction[i]){
            res[item].insert(i+1);
        }
    }
    return res;
}

void ECLAT(vector<pair<string, set<int>>> items, set<int> prefixTids, vector<string> prefix, int minSup,
    vector<pair<vector<string>, int>>& results){
    for(size_t i = 0; i < items.size(); ++i){
        const string item = items[i].first;
        const set<int> itemTids = items[i].second;

        set<int> newTids;
        if (prefix.empty()){
            newTids = itemTids; 
        } else {
            auto it1 = prefixTids.begin();
            auto it2 = itemTids.begin();
            
            while (it1 != prefixTids.end() && it2 != itemTids.end()){
                if (*it1 < *it2) {
                    ++it1;
                } else if (*it2 < *it1) {
                    ++it2;
                } else {
                    newTids.insert(*it1);
                    ++it1;
                    ++it2;
                }
            }
        }
        
        int support = newTids.size();
        if (support >= minSup){
            vector<string> newItemset = prefix;
            newItemset.push_back(item);
            
            results.push_back({newItemset, support});
            
            vector<pair<string, set<int>>> newItems;
            for (size_t j = i + 1; j < items.size(); ++j){
                newItems.push_back(items[j]);
            }
            
            if (!newItems.empty()){
                ECLAT(newItems, newTids, newItemset, minSup, results);
            }
        }
    }
}

bool comp(pair<string,set<int>> a, pair<string, set<int>> b){
    return a.second.size() < b.second.size();
}

signed main(){
    vector<vector<string>> transaction ={
        {"f", "a", "c", "d", "g", "i", "m", "p"},
        {"a", "b", "c", "f", "l", "m", "o"},
        {"b", "f", "h", "j", "o"},
        {"b", "c", "k", "s", "p"},
        {"a", "f", "c", "e", "l", "p", "m", "n"}
    };
    
    map<string, set<int>> convert_transaction= convert(transaction);
    // for(pair<string, set<int>> x : convert_transaction){
    //     cout << x.first << ": ";
    //     for(int i:x.second){
    //         cout << i << " ";
    //     }
    //     cout << endl;
    // }
    int minSup=3;

    // loc cac itemset co sup < minSup
    vector<pair<string, set<int>>> itemSet;
    for(pair<string, set<int>> x : convert_transaction){
        if((int)x.second.size() >= minSup){
            itemSet.push_back({x.first, x.second});
        }
    }

    //sap xep cac itemset theo so luong TID trong cac itemset tang dan
    //vd: milk:  1 2 5 4
    //    butter 1 4 3
    //    bread  1 2
    sort(itemSet.begin(), itemSet.end(), comp);

    // luu kq
    vector<pair<vector<string>,int>> ans;
    ECLAT(itemSet, set<int>{}, vector<string>{}, minSup, ans);

    cout << "cac tap pho bien:" << endl;
    for(pair<vector<string>, int> x : ans){
        //item
        for(string y : x.first){
            cout << y << " ";
        }
        //sup
        cout << ": " << x.second << endl;
    }
}